src/coreclr/jit/flowgraph.cpp

```diff
@@ -48,6 +48,10 @@ static bool blockNeedsGCPoll(BasicBlock* block)
                         blockMayNeedGCPoll = true;
                     }
                 }
+                else if (tree->OperGet() == GT_GCPOLL)
+                {
+                    blockMayNeedGCPoll = true;
+                }
```

Comment:
This seems like the GC poll will not be inserted in the place the programmer wrote the call. Does it matter?

Is there a reason to make this a special intrinsic instead of a managed call to a function that checks the flag and that can get inlined in the right spot?

Reply:
the current logic allows us to remove it if there are other safe points in the current block (e.g. a pinvoke without SGCT)

>  call to a function that checks the flag

which is how it looked before this change and caused perf regressions due to TLS access while we can optimize it with just an address of compile-time known global field "poll might be needed"

Reply:
> It will be at the end of some internal BasicBlock which will depend on number of factors (like QMARKs or other expansions). So it will probably be hard for the user to reason about the exact place this GC poll is going to happen.
But it sounds like we are ok with that.

If that is an issue we can easily re-use the helpercallexpansion, we probably can actually do it as a clean up for current code

> I suppose an alternative could have been to provide a way to access g_TrapReturningThreads from the managed code.

Yeah, it's an option. But we'll have to introduce a InternalCall for actual poll and an intrinsic to get the `g_TrapReturningThreads`, seems like the amount of changes will be the same?

Reply:
> We probably can re-use helpercallexpansion here to make it a bit more precise if needed (or as a clean up)

I like that idea the most. We could represent the poll as just a `CORINFO_HELP_POLL_GC` helper call and then do the helper-expansion in optimized codegen to insert the fast check around it.

I think that would also allow us to remove `OMF_NEEDS_GCPOLLS`, `BBF_HAS_SUPPRESSGC_CALL` and `BBF_NEEDS_GCPOLL`.

(No need from my side to do that in this PR.)

Reply:
> I like that idea the most. We could represent the poll as just a CORINFO_HELP_POLL_GC helper call and then do the helper-expansion in optimized codegen to insert the fast check around it.

agree, I can handle it in a follow up

Reply:
Thanks all. I'm assuming this is currently okay for now and there is a follow-up on the JIT team to do something more.

Reply:
> the current logic allows us to remove it if there are other safe points in the current block (e.g. a pinvoke without SGCT),

I think that it is the desired behavior. I do not think we need any cleanups on top what's in this PR.

Reply:
I think the desired behavior would be that this optimization kicks in in a predictable fashion. It is completely unpredictable when this optimization will kick in because basic blocks are an internal JIT concept.

Reply:
I would still like to see this clean up happen. I do not think we need additional new nodes to represent this concept, and I think the current flag based representation used e.g. by tailcalls is worse than representing it vis IR.

But I would be fine with doing the same "skip multiple GC polls in the same basic block" optimization if @jkotas thinks this is important.

Reply:
> I think the desired behavior would be that this optimization kicks in in a predictable fashion.

I was commenting on the original question whether the GC poll needs to be in the exact spot where the programmer wrote the call.

The predictable behavior here means that the explicit GC poll (or some other opportunity for GC suspension) happens somewhere along the path. It is ok for JIT to move the poll a bit later, and it is desirable optimization to fold it with other GC polls. The observable behavior is that the thread suspension does not take too long, the exact spot where the thread gets suspended does not matter.

If you would like to clean up the internal JIT representation, that's fine.

Reply:
> It is ok for JIT to move the poll a bit later

The current implementation can also do the poll earlier (at the beginning of the basic block) in various cases -- not optimizing, R2R codegen, code is considered cold or code is inside a switch case. If the poll is only there in case of inlining into some tight loop then I suppose that's ok as well.

> desirable optimization to fold it with other GC polls

I understand that. But if we want to claim we do this optimization we would need to consider larger units, like cycles or regions of basic blocks instead. Considering just single basic blocks is quite fragile and has large probability to break because of IL or JIT changes.